lxd/operations.go

```diff
@@ -44,6 +47,69 @@ var operationWebsocket = APIEndpoint{
 	Get: APIEndpointAction{Handler: operationWebsocketGet, AllowUntrusted: true},
 }
 
+func waitForOperations(s *state.State, force chan bool) {
+	timeout := time.After(10 * time.Minute)
+	tick := time.Tick(time.Second)
+	logTick := time.Tick(time.Minute)
+
+	for {
+		<-tick
+
+		// Get all the operations
+		operations.Lock()
+		ops := operations.Operations()
+		operations.Unlock()
+
+		runningOps := 0
+
+		for _, op := range ops {
+			if op.Status() == api.Running {
+				runningOps++
+			}
+		}
+		// No more running operations left. Exit function.
+		if runningOps == 0 {
+			logger.Info("All running operations finished, shutting down")
+			return
+		}
+
+		execConsoleOps := 0
+
+		for _, op := range ops {
+			opType := op.Type()
+
+			if opType == db.OperationCommandExec || opType == db.OperationConsoleShow {
+				execConsoleOps++
+			}
+
+			_, opAPI, _ := op.Render()
+
+			if opAPI.MayCancel {
+				op.Cancel()
+			}
+		}
+
+		select {
+		case <-timeout:
+			// We wait up to 10 minutes for exec/console operations to finish.
+			// If there are still running operations, we shut down the instances
+			// which will terminate the operations.
+			if execConsoleOps > 0 {
+				logger.Info("Shutdown timeout reached, shutting down instances")
+				instancesShutdown(s)
+			}
+
+		case <-logTick:
+			// Print log message every minute.
+			logger.Infof("Waiting for %d operation(s) to finish", runningOps)
+		case <-force:
+			// Return here, and ignore any running operations.
```

Comment:
@monstermunchkin maybe log an error here saying that we are exiting without cleanly finishing all operations because they took too long.